Definition validation, no export customization, base URL change

glasser opened the pull request:
This PR contains three changes. See individual commits for details.

glasser committed:
Processors must validate directive definitions

Following discussion with @pcmanus and @martijnwalraven, we've decided that
compliant processors should expect directives to be defined with arguments
matching what the specification expects. To define a directive with an argument
whose type or default value differs from the specification is now an error that
should be detected, rather than something that could lead to surprising
processing (eg, a processor treating `@core__export` as equivalent to
`@core__export(export: true)` even if a specific schema defines the
directive as `directive @core__export(export: Boolean! = false)`.

Specifically, what we validate is:
- All expected arguments are provided with the correct types and default values
- No additional arguments are provided
- The `repeated` flag is as expected
- The list of locations is as expected

We do not validate:
- Order of arguments or locations
- Description string on the directive or any arguments
- Directives applied to the argument definitions

This should be sufficient to ensure that, once the definitions have been
validated, later processing code can assume that directive arguments have the
expected types and the semantics of omitting out an argument is well-defined.

We remove from the spec the concept of the extensibility of directives (and
other types). We still think this is a good idea, but we'd like to wait until we
have a use case for it. It makes defining this validation simpler if it doesn't
have to include sometimes allowing additional arguments.

Include a top-level section with both directive definitions that can be easily
copied into core schemas. Update most examples to refer to this section instead
of trying to define the directives.

glasser commented:
Preview available at https://apollo-specs.github.io/core/glasser/directive-definition-validation/

Might not be a bad idea to skim the whole spec to make sure I didn't leave anything in that refers to the stuff that has changed.

glasser commented on core.spec.md:
```diff
@@ -380,15 +424,6 @@ Schema documents MUST reference a feature version which supports all the schema
 If a processor chooses to activate support for a feature, the processor MUST activate an implementation which can [satisfy](#sec-Satisfaction) the version required by the document.
 
 
-# Extensibility
```
I feel a little bad about entirely removing this but I think we're right that we should wait until we have a use case for keeping this. If we do put this back we'll want to change the bootstrapping algorithm to check that the argument list and location lists have subset relationships rather than equality, etc.

pcmanus commented on core.spec.md:
````diff
@@ -138,11 +138,54 @@ type SomeType {
 directive @example on FIELD_DEFINITION
 ```
 
+# Directive Definitions
````
When reading that section, I feel it's not 100% clear whether the section refers only exclusively to the `@core` spec definitions, or if it define constraints for nay "feature". I believe we mean the later, and the section does mention "other core features", but I feel this can be clarified.  In particular, the spec allows to omit the `@core_export` definition in one's document, but it's never clearly said that this is a general allowance (to leave unused directives out). And do we actually want to allow that in general? I think we do, but if so, we should have a clear list of what you can leave out in general. We could say that you can leave full `Definition` (in the sense of http://spec.graphql.org/draft/#Definition) out, but that you cannot leave a sub-part of a definition out, nor modify it.

glasser replied:
Removed export in next section. Clarified that this is about `@core` itself but is also a model for other features.

pcmanus commented on bad-non-unique-prefix-multi.graphql:
```diff
@@ -6,5 +6,5 @@ schema
 
-directive @core(feature: String!, as: String, export: Boolean)
-  repeatable on SCHEMA
+# Definitions for this spec's directives have been elided; the definitions
```
Assuming we agree that the answer to the "is it legal to leave unused directives definitions out?" in my other comment is "yes", then I'd suggests removing those comments and replace them by the actual definitions.  This avoid any ambiguity, and we have very few examples actually using `@core_export`, so this won't make the document longer (the comment here is 2 lines but you can easily fit the `@code` directive definition on 2 lines, if not just a single line).

glasser replied:
Now that `@core__export` is gone, this suggestion is easy.

pcmanus commented on core.spec.md:
```diff
@@ -320,7 +364,7 @@ The prefix MUST NOT be elided within documentation; definitions of schema elemen
 Feature specs MUST prefix the following schema elements:
   - the names of any object types, interfaces, unions, enums, or input object types defined by the feature
   - the names of any fields the spec introduces on *foreign types* defined in a different feature
```
Not sure we should keep this bullet and the next one (anything that mentions a "foreign" spec basically), because 1) I don't think the spec is currently clear on how such a thing should work/be valid and 2) I'm not sure how to even make that work with the whole "you can just copy-paste" definitions.

Tbc, I'm sure we'll have needs for combining specs somewhat at some point, but it'll be easier to think that through when we have concrete use cases, and it's easier to add new capability later than to remove them.

If we do remove this, we should also update the example of that section.

glasser replied:
Removed these lines.

glasser committed:
Remove customizability of export from spec

There's some disagreement on the team as to whether this belongs in Core Schema
itself or something separate. The only planned short-term use case for these
features is the "inaccessible fields" proposal and there is disagreement as to
whether `@core__export` is actually appropriate for that use case.

It is very likely we will add something like this back sometime but let's keep
it out of v0.1 for now.

It turns out that it now is easy for us to entirely remove the term "export"
from the spec and stick with the newer "API" term.

It also means that actually declaring the directive in every example is
non-burdensome so we do that.

Also remove references to extensions of foreign types since we don't have great
examples/use cases for that right now.

glasser committed:
Change lib.apollo.dev to specs.apollo.dev

Neither URL works yet, but we think this is more clear than "lib" which suggests
library code.

glasser commented:
@pcmanus @martijnwalraven This PR now contains three separate commits with commit messages. The first commit is the same that you reviewed already; the other two have been added based on our discussions.

pcmanus commented:
Had a pass over the last version and my only remark is that I think the foreign fields (`B__foreignField` and `B__field`) of example 7 in section 6.1 should be removed now that the spec don't talk about extensibility.  Especially with section 4 saying that spec should essentially provide definitions for copy-pasting, this kind of example feels confusing.

glasser committed:
Remove more references to extensibility
